FWCore/Framework/src/EventProcessor.cc

```diff
@@ -2256,6 +2260,7 @@ namespace edm {
                 // If appropriate, this will also start the next run.
                 endRunAsync(streamRunStatus_[iStreamIndex], iTask);
               }
+              status->noMoreEventsInLumi();
```

Comment:
Just thinking out loud for a possible (logical) data race.

The `noMoreEventsInLumi()` sets the `LuminosityBlockProcessingStatus::nStreamsStillProcessingLumi_`, is called after the `startNextLumiOrEndRun()`.

**If** some other thread would call the `handleNextEventForStreamAsync()` for the same stream before this thread calls `noMoreEventsInLumi()`, but after it called `startNextLumiOrEndRun()`, the other thread would call `streamEndLumiAsync()`, where it would see `status->streamFinishedLumi()` returning `true` (because the status->nStreamsStillProcessingLumi_` would still be zero at that point).

It is not clear to me though if the situation above could happen. The `handleNextEventForStreamAsync()` is called from `beginLumiAsync()`, from a task queued in `streamQueues_[iStreamIndex]`, and only after the task paused the `streamQueues_[iStreamIndex]`.

I might well be missing something in either direction.